vernac/vernacentries.ml

```diff
@@ -2301,9 +2378,17 @@ let rec interp_expr ?proof ~atts ~st c : Proof_global.t option =
     vernac_inductive ~atts cum priv finite l;
     pstate
   | VernacFixpoint (discharge, l) ->
-    with_def_attributes ~atts vernac_fixpoint ~pstate discharge l
+    let opens = List.exists (fun ((_,_,_,_,p),_) -> Option.is_empty p) l in
+    if opens then
+      with_open_proof_simple ~pstate (with_def_attributes ~atts vernac_fixpoint_interactive discharge l)
+    else
+      (with_def_attributes ~atts vernac_fixpoint discharge l; pstate)
```

Comment:
This improves little the previous situation, at the static level still the type of the interpretation of `VernacFixpoint` is `Proof.t option`.

Reply:
In fact this PR in its current form may introduce a serious problem, as it is not clear it is possible to provide static guarantees anymore. IMHO the path forward if you want to provide a static guarantee w.r.t. proof opening is:

- merge #10050
- merge to `master` all the bits specific to #10215 except the removal of `MaybeOpenProof`
- merge #10230 which provides static guarantees to all vernacs [except `Qed`, but that's a different story]
- _now_, which strong typing guarantees, really remove `MaybeOpenProof`, which will maybe imply some refactoring in `vernac_expr`.

Reply:
> if you want to provide a static guarantee w.r.t. proof opening

In case it is not clear, I don't want (nor claim) to provided a static guarantee of anything in this PR.

I just want to avoid making it legal to declare a new vernacular without knowing if it opens a proof, that means no `![maybe_open_proof]` annotation in coqpp and not "typing" `VtMaybeOpenProof`.
Of course, given that the type of the main loop of the interpreter is still `... ->proof option` there is static guarantee as you put it that someone sneaks in vernac_expr a new command that really needs such an imprecise type.

While I can keep an eye on master and void a new command with that "type" gets introduced, and as I did in this PR I can clean existing code when the type is not really needed, I can for sure not prevent plugin writers from using the brand new `![maybe_open_proof]`. This is, again, sorry for repeating myself, why #10050 was not OK to me.

> merge #10050

This pr was *closed by its author in favor of this one*.

> merge #10230 which provides static guarantees to all vernacs [except `Qed`, but that's a different story]

This PR is a draft that does not pass CI.